File: src/tools/message_queue.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { GraphQLClient } from 'graphql-request';
import { z } from 'zod';
import {
  MessageQueuesGetDocument,
  MessageQueueGetDocument,
  MessageQueueCreateDocument,
  MessageQueueModifyDocument,
  MessageQueueDeleteDocument,
  MessageQueuePlansGetDocument,
  MessageQueueVersionsGetDocument,
  MessageQueueUserCredentialsGetDocument,
  type MessageQueuesGetQuery,
  type MessageQueueGetQuery,
  type MessageQueueCreateMutation,
  type MessageQueueModifyMutation,
  type MessageQueuePlansGetQuery,
  type MessageQueueVersionsGetQuery,
  type MessageQueueUserCredentialsGetQuery,
} from '../generated/graphql.js';

const AllowListEntry = z.object({
  ip: z.string(),
  state: z.enum(['PRESENT', 'ABSENT']),
});

const ExternalConnectionPort = z.object({
  externalPort: z
    .number()
    .optional()
    .describe(
      'Omit when creating a new port — the platform assigns the external port automatically. Required when modifying or removing an existing port so the backend knows which port to target.',
    ),
  internalPort: z.number().optional(),
  protocol: z.enum(['TCP', 'UDP']),
  state: z.enum(['PRESENT', 'ABSENT']),
  allowList: z.array(AllowListEntry),
});

const ExternalConnection = z.object({
  sharedIp: z.boolean(),
  state: z.enum(['PRESENT', 'ABSENT']),
  ports: z.array(ExternalConnectionPort),
});

export function registerMessageQueueTools(server: McpServer, client: GraphQLClient): void {
  server.registerTool(
    'nexaa_message_queue_list',
    { description: 'List all message queues' },
    async () => {
      const data = await client.request<MessageQueuesGetQuery>(MessageQueuesGetDocument);
      return {
        content: [{ type: 'text', text: JSON.stringify(data.messageQueues, null, 2) }],
      };
    },
  );

  server.registerTool(
    'nexaa_message_queue_get',
    {
      description: 'Get a message queue by name and namespace',
      inputSchema: {
        name: z.string(),
        namespace: z.string(),
      },
    },
    async ({ name, namespace }) => {
      const data = await client.request<MessageQueueGetQuery>(MessageQueueGetDocument, {
        messageQueueInput: { name, namespace },
      });
      return {
        content: [{ type: 'text', text: JSON.stringify(data.messageQueue, null, 2) }],
      };
    },
  );

  server.registerTool(
    'nexaa_message_queue_create',
    {
      description:
        'Create a message queue. Use nexaa_message_queue_list_plans and nexaa_message_queue_list_versions to discover valid plan IDs and spec values.',
      inputSchema: {
        name: z.string(),
        namespace: z.string(),
        plan: z.string().describe('Plan ID from nexaa_message_queue_list_plans'),
        spec: z.object({
          type: z.string().describe('e.g. rabbitmq'),
          version: z.string().describe('e.g. 3'),
          patchLevelVersion: z.string().optional(),
        }),
        allowList: z.array(AllowListEntry).default([]),
        externalConnection: ExternalConnection.optional(),
      },
    },
    async (input) => {
      const data = await client.request<MessageQueueCreateMutation>(MessageQueueCreateDocument, {
        messageQueueInput: input,
      });
      return {
        content: [{ type: 'text', text: JSON.stringify(data.messageQueueCreate, null, 2) }],
      };
    },
  );

  server.registerTool(
    'nexaa_message_queue_modify',
    {
      description: "Modify a message queue's allowlist or external connection",
      inputSchema: {
        name: z.string(),
        namespace: z.string(),
        allowList: z.array(AllowListEntry).default([]),
        externalConnection: ExternalConnection.optional(),
      },
    },
    async (input) => {
      const data = await client.request<MessageQueueModifyMutation>(MessageQueueModifyDocument, {
        messageQueueInput: input,
      });
      return {
        content: [{ type: 'text', text: JSON.stringify(data.messageQueueModify, null, 2) }],
      };
    },
  );

  server.registerTool(
    'nexaa_message_queue_delete',
    {
      description:
        'Delete a message queue. WARNING: This is irreversible. Requires confirm: true.',
      inputSchema: {
        name: z.string(),
        namespace: z.string(),
        confirm: z.boolean().describe('Must be set to true to confirm deletion'),
      },
    },
    async ({ name, namespace, confirm }) => {
      if (!confirm) {
        return {
          content: [
            {
              type: 'text',
              text: 'Deletion cancelled: confirm must be true to delete a message queue.',
            },
          ],
        };
      }
      await client.request(MessageQueueDeleteDocument, { messageQueueInput: { name, namespace } });
      return {
        content: [
          {
            type: 'text',
            text: `Message queue "${name}" deleted from namespace "${namespace}".`,
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_message_queue_list_plans',
    {
      description:
        'List available message queue plans with CPU, memory, storage, and pricing information',
    },
    async () => {
      const data = await client.request<MessageQueuePlansGetQuery>(MessageQueuePlansGetDocument);
      return {
        content: [{ type: 'text', text: JSON.stringify(data.messageQueuePlans, null, 2) }],
      };
    },
  );

  server.registerTool(
    'nexaa_message_queue_list_versions',
    {
      description: 'List supported message queue engine types and versions',
    },
    async () => {
      const data = await client.request<MessageQueueVersionsGetQuery>(
        MessageQueueVersionsGetDocument,
      );
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.messageQueueVersions, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_message_queue_get_credentials',
    {
      description:
        'Retrieve credentials (DSN, password) for a message queue user. Treat the result as sensitive.',
      inputSchema: {
        name: z.string().describe('Message queue name'),
        namespace: z.string(),
        username: z.string(),
      },
    },
    async ({ name, namespace, username }) => {
      const data = await client.request<MessageQueueUserCredentialsGetQuery>(
        MessageQueueUserCredentialsGetDocument,
        {
          messageQueueInput: { name, namespace },
          username,
        },
      );
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.messageQueueUserCredentials, null, 2),
          },
        ],
      };
    },
  );
}

File: tools/message_queue.go
package tools

import (
	"context"
	"fmt"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"github.com/nexaa-cloud/nexaa-cli/api"
)

var allowListItemSchema = map[string]any{
	"type": "object",
	"properties": map[string]any{
		"ip":    map[string]any{"type": "string"},
		"state": map[string]any{"type": "string", "enum": []string{"PRESENT", "ABSENT"}},
	},
	"required": []string{"ip", "state"},
}

func toAllowList(args []allowListArg) []api.AllowListInput {
	out := make([]api.AllowListInput, len(args))
	for i, a := range args {
		out[i] = api.AllowListInput{Ip: a.IP, State: api.State(a.State)}
	}
	return out
}

func RegisterMessageQueue(s *server.MCPServer) {
	s.AddTool(
		mcp.NewTool("nexaa_message_queue_list",
			mcp.WithDescription("List all message queues"),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			result, err := withAPI(func(c *api.Client) ([]api.MessageQueueResult, error) {
				return c.MessageQueueList()
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_message_queue_get",
			mcp.WithDescription("Get a message queue by name and namespace"),
			mcp.WithString("name", mcp.Required()),
			mcp.WithString("namespace", mcp.Required()),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			result, err := withAPI(func(c *api.Client) (api.MessageQueueResult, error) {
				return c.MessageQueueGet(api.MessageQueueResourceInput{
					Name:      req.GetString("name", ""),
					Namespace: req.GetString("namespace", ""),
				})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_message_queue_create",
			mcp.WithDescription("Create a message queue. Use nexaa_message_queue_list_plans and nexaa_message_queue_list_versions to discover valid plan IDs and spec values."),
			mcp.WithString("name", mcp.Required()),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithString("plan", mcp.Required(), mcp.Description("Plan ID from nexaa_message_queue_list_plans")),
			mcp.WithObject("spec", mcp.Required(), mcp.Description("Message queue engine spec"),
				mcp.Properties(map[string]any{
					"type":    map[string]any{"type": "string", "description": "e.g. rabbitmq"},
					"version": map[string]any{"type": "string", "description": "e.g. 3"},
				}),
			),
			mcp.WithArray("allowList", mcp.Items(allowListItemSchema)),
			mcp.WithObject("externalConnection"),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			var args struct {
				Name               string         `json:"name"`
				Namespace          string         `json:"namespace"`
				Plan               string         `json:"plan"`
				Spec               struct {
					Type    string `json:"type"`
					Version string `json:"version"`
				} `json:"spec"`
				AllowList          []allowListArg `json:"allowList"`
				ExternalConnection *extConnArg    `json:"externalConnection"`
			}
			if err := req.BindArguments(&args); err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			result, err := withAPI(func(c *api.Client) (api.MessageQueueResult, error) {
				return c.MessageQueueCreate(api.MessageQueueCreateInput{
					Name:               args.Name,
					Namespace:          args.Namespace,
					Plan:               args.Plan,
					Spec:               api.MessageQueueSpecInput{Type: args.Spec.Type, Version: args.Spec.Version},
					AllowList:          toAllowList(args.AllowList),
					ExternalConnection: toExtConn(args.ExternalConnection),
				})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_message_queue_modify",
			mcp.WithDescription("Modify a message queue's allowlist or external connection"),
			mcp.WithString("name", mcp.Required()),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithArray("allowList", mcp.Items(allowListItemSchema)),
			mcp.WithObject("externalConnection"),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			var args struct {
				Name               string         `json:"name"`
				Namespace          string         `json:"namespace"`
				AllowList          []allowListArg `json:"allowList"`
				ExternalConnection *extConnArg    `json:"externalConnection"`
			}
			if err := req.BindArguments(&args); err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			result, err := withAPI(func(c *api.Client) (api.MessageQueueResult, error) {
				return c.MessageQueueModify(api.MessageQueueModifyInput{
					Name:               args.Name,
					Namespace:          args.Namespace,
					AllowList:          toAllowList(args.AllowList),
					ExternalConnection: toExtConn(args.ExternalConnection),
				})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_message_queue_delete",
			mcp.WithDescription("Delete a message queue. WARNING: This is irreversible. Requires confirm: true."),
			mcp.WithString("name", mcp.Required()),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithBoolean("confirm", mcp.Required(), mcp.Description("Must be true to confirm deletion")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			name := req.GetString("name", "")
			ns := req.GetString("namespace", "")
			if ok, cancelled := confirmDelete(req, "message queue"); !ok {
				return cancelled, nil
			}
			_, err := withAPI(func(c *api.Client) (bool, error) {
				return c.MessageQueueDelete(api.MessageQueueResourceInput{Name: name, Namespace: ns})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return mcp.NewToolResultText(fmt.Sprintf("Message queue %q deleted from namespace %q.", name, ns)), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_message_queue_list_plans",
			mcp.WithDescription("List available message queue plans with CPU, memory, storage, and pricing information"),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			result, err := withAPI(func(c *api.Client) ([]api.MessageQueuePlanResult, error) {
				return c.MessageQueuePlans()
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_message_queue_list_versions",
			mcp.WithDescription("List supported message queue engine types and versions"),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			result, err := withAPI(func(c *api.Client) ([]api.MessageQueueVersionResult, error) {
				return c.MessageQueueVersions()
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_message_queue_get_credentials",
			mcp.WithDescription("Retrieve credentials (DSN, password) for a message queue user. Treat the result as sensitive."),
			mcp.WithString("name", mcp.Required(), mcp.Description("Message queue name")),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithString("username", mcp.Required()),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			result, err := withAPI(func(c *api.Client) (api.MessageQueueUserCredentialsResult, error) {
				return c.MessageQueueAdminCredentials(
					api.MessageQueueResourceInput{
						Name:      req.GetString("name", ""),
						Namespace: req.GetString("namespace", ""),
					},
					req.GetString("username", ""),
				)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)
}
